src/inlinestrings.jl

```diff
@@ -0,0 +1,770 @@
+import Base: ==
+
+export InlineString, InlineStringType
+
+abstract type InlineString <: AbstractString end
+
+for sz in (1, 4, 8, 16, 32, 64, 128, 256)
+    nm = Symbol(:InlineString, max(1, sz - 1))
+    @eval begin
+        primitive type $nm <: InlineString $(sz * 8) end
+        export $nm
+    end
+end
+
+_bswap(x::T) where {T <: InlineString} = T === InlineString1 ? x : Base.bswap_int(x)
+
+const InlineStrings = Union{InlineString1,
```

Comment:
do we really need that many options? Also in CSV.jl how would be the width of the `InlineString*` chosen? Automatically?

Reply:
Automatically. In single threaded, we do smallest inline string type that fits first string we find; if we "overflow", then we promote to wider inline string type, like Int => Float promotion. Promoting like this is pretty cheap, since it's a lossless conversion. For multithreaded, we track the largest string size we find per column during type sampling, and will use that to choose inline string type.